Expectations model

trew opened the pull request:
Added initial model for expectations. 

This model does not accept weeks as a timedelta. Trying to do so gave me headache, and lots of ugly solutions were forced, so I figured weeks must be handled clientside. That should be easy because a week is always 7 days.

trew committed:
added expectations model with a few associated functions

trew committed:
using relativedelta instead of timedelta because timedelta couldn't handle months correctly.
minor bugfixes as well.

trew committed:
added error messages and static methods to retrieve expectation unit from string or int

trew committed:
made get_timedelta static. check_expectation() works and returns a dict of errors instead of true/false. empty dict would represent true. also made simple constructor for easy testing purposes

martinmelin commented on monlog/log/models.py:
```diff
@@ -71,3 +84,126 @@ class Label(Filter):
         return self.label_name
 
+
+class Expectation(Label):
```
Should Expectation subclass Label or Filter?

trew replied:
At first, Expectation was subclass to Filter, but then I figured, Expectations is created by users(foreignkey to user), and they also have a name(label_name). Both are present in Label.

trew replied:
Using Filter instead. Django created a key relation in the database which we don't want. Instead I added name and user fields to the Expectation model.

martinmelin commented on monlog/log/models.py:
```diff
@@ -71,3 +84,126 @@ class Label(Filter):
+    # +- amount of unit
+    # example: '+- 10 minute'
+    tolerance_amount = models.IntegerField()
+    tolerance_unit = models.IntegerField()
```
Can we set choices=EXPECTATION_UNITS here and for repeat_unit?

trew replied:
Heyo, expecation_units is gone. It was ugly anyway.

martinmelin commented on monlog/log/models.py:
```diff
@@ -71,3 +84,126 @@ class Label(Filter):
+        self.repeat_unit = repeat_unit
+        self.least_amount_of_hits = least_amount_of_hits
+        self.label_name = label_name
+        self.apply_tolerance()
```
I think we should do this in check_expectation instead. I think this will break otherwise if you change the deadline of an already instantiated Expectation.

trew replied:
As I wrote in the commit message, this constructor is quite temporary overall. It's only been used recently to simplify testing. But I agree, calling apply_tolerance() can absolutely be done in check_expectation().

trew replied:
Apply_tolerance is now being called in check_expectation, but is instead modifying a querydict. It makes more sense to do so instead of modifiying the original query_string.

martinmelin commented on monlog/log/models.py:
```diff
@@ -71,3 +84,126 @@ class Label(Filter):
+
+        qs = QueryDict(self.query_string, mutable=True)
+        qs['datetime__gte'] = startdate.isoformat().replace('T',' ')
+        qs['datetime__lte'] = enddate.isoformat().replace('T',' ')
```
This .replace('T') nonsense is crap :-) Remind me to look into convincing the validation to accept valid ISO 8601 (i.e. with T)

trew replied:
I agree. Django raises this:

```
ValidationError: [u'Enter a valid date/time in YYYY-MM-DD HH:MM[:ss[.uuuuuu]] format.']
```

trew replied:
This is apparently fixed in Django 1.4 that was released 2 days ago. I looked in the django 1.3 code around the exception and it wasn't pretty. New version uses a more flexible regexp to parse the datetime. See django.utils.dateparse.

martinmelin commented on monlog/log/models.py:
```diff
@@ -71,3 +84,126 @@ class Label(Filter):
+            print "Error: Couldn't get next deadline."
+            return None
+
+    @staticmethod
```
What do you think of making .timedelta a property of the instance instead? That way we could just say self.timedelta instead of passing in self.repeat_*

http://docs.python.org/library/functions.html#property

martinmelin commented:
Holy dragbegäran, batman!

This looks great, seems like we've come a long way already. I added some line comments because honestly, what else am I going to do over lunch?

trew committed:
added relativedeltafield class to be used in expectation model

trew committed:
apply_tolerance() modifies a querydict instead of the original query_string. makes more sense. is being called in check_expectation().
created a RelativedeltaField in order to make repeat and tolerance more intuitive. a relativedelta object is saved as a string in database.

trew committed:
adjusted some comments and removed some imports

trew committed:
removed ugly hacks for django 1.3. monlog now require django 1.4

trew committed:
renamed variable

trew committed:
retrieved new django manage.py

trew committed:
check_expectation() returns queryset as well.
